<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.52">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Tong Zeng">
<meta name="author" content="Longfeng Wu">
<meta name="author" content="Sarah Bratt">
<meta name="author" content="Daniel E Acuna">
<meta name="dcterms.date" content="2020-01-24">
<meta name="keywords" content="DataRank, Scientiﬁc dataset, Dataset impact, Citation network">
<meta name="description" content="DataRank - an algorithm to properly track and assign credit to datasets in citation network.">

<title>Assigning credit to scientific datasets using article citation networks – Tong Zeng</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=GTM-TKJ3X4RT"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'GTM-TKJ3X4RT', { 'anonymize_ip': false});
</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
<meta name="citation_title" content="Assigning credit to scientific datasets using article citation networks">
<meta name="citation_abstract" content="A citation is a well-established mechanism for connecting scientiﬁc artifacts. Citation networks are used by citation analysis for a variety of reasons, prominently to give credit to scientists’ work. However, because of current citation practices, scientists tend to cite only publications, leaving out other types of artifacts such as datasets. Datasets then do not get appropriate credit even though they are increasingly reused and experimented with. We develop a network ﬂow measure, called DataRank, aimed at solving this gap. DataRank assigns a relative value to each node in the network based on how citations ﬂow through the graph, differentiating publication and dataset ﬂow rates. We evaluate the quality of DataRank by estimating its accuracy at predicting the usage of real datasets: web visits to GenBank and downloads of Figshare datasets. We show that DataRank is better at predicting this usage compared to alternatives while offering additional interpretable outcomes. We discuss improvements to citation behavior and algorithms to properly track and assign credit to datasets.
">
<meta name="citation_keywords" content="DataRank,Scientiﬁc dataset,Dataset impact,Citation network">
<meta name="citation_author" content="Tong Zeng">
<meta name="citation_author" content="Longfeng Wu">
<meta name="citation_author" content="Sarah Bratt">
<meta name="citation_author" content="Daniel E Acuna">
<meta name="citation_publication_date" content="2020-05-01">
<meta name="citation_cover_date" content="2020-05-01">
<meta name="citation_year" content="2020">
<meta name="citation_online_date" content="2020-01-24">
<meta name="citation_fulltext_html_url" content="https://www.sciencedirect.com/science/article/pii/S1751157719301841">
<meta name="citation_pdf_url" content="https://www.sciencedirect.com/science/article/am/pii/S1751157719301841">
<meta name="citation_issue" content="2">
<meta name="citation_doi" content="10.1016/j.joi.2020.101013">
<meta name="citation_issn" content="1751-1577">
<meta name="citation_volume" content="14">
<meta name="citation_language" content="en">
<meta name="citation_firstpage" content="101013">
<meta name="citation_journal_title" content="Journal of Informetrics">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Tong Zeng</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../#news"> 
<span class="menu-text">News</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../#publications"> 
<span class="menu-text">Publications</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blogs.html"> 
<span class="menu-text">Blogs</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/tong-zeng/tong-zeng.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/zeng_tong"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:tongzeng@vt.edu"> <i class="bi bi-envelope" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Assigning credit to scientific datasets using article citation networks</h1>
                  <div>
        <div class="description">
          DataRank - an algorithm to properly track and assign credit to datasets in citation network.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">Science of Science</div>
                <div class="quarto-category">Network Science</div>
                <div class="quarto-category">DATARANK</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Tong Zeng <a href="mailto:tongzeng@vt.edu," class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> </p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Information Management, Nanjing University, Nanjing 210023, China
            </p>
          <p class="affiliation">
              School of Information Studies, Syracuse University, Syracuse, NY 13244, USA
            </p>
        </div>
      <div class="quarto-title-meta-contents">
      <p class="author">Longfeng Wu </p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Information Studies, Syracuse University, Syracuse, NY 13244, USA
            </p>
        </div>
      <div class="quarto-title-meta-contents">
      <p class="author">Sarah Bratt </p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Information Studies, Syracuse University, Syracuse, NY 13244, USA
            </p>
        </div>
      <div class="quarto-title-meta-contents">
      <p class="author">Daniel E Acuna <a href="mailto:daniel.acuna@colorado.edu," class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-7765-1595" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              School of Information Studies, Syracuse University, Syracuse, NY 13244, USA
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 24, 2020</p>
      </div>
    </div>
    
      
    </div>
    
  <div>
    <div class="abstract">
      <div class="block-title">Abstract</div>
      <p>A citation is a well-established mechanism for connecting scientiﬁc artifacts. Citation networks are used by citation analysis for a variety of reasons, prominently to give credit to scientists’ work. However, because of current citation practices, scientists tend to cite only publications, leaving out other types of artifacts such as datasets. Datasets then do not get appropriate credit even though they are increasingly reused and experimented with. We develop a network ﬂow measure, called DataRank, aimed at solving this gap. DataRank assigns a relative value to each node in the network based on how citations ﬂow through the graph, differentiating publication and dataset ﬂow rates. We evaluate the quality of DataRank by estimating its accuracy at predicting the usage of real datasets: web visits to GenBank and downloads of Figshare datasets. We show that DataRank is better at predicting this usage compared to alternatives while offering additional interpretable outcomes. We discuss improvements to citation behavior and algorithms to properly track and assign credit to datasets.</p>
    </div>
  </div>

  <div>
    <div class="keywords">
      <div class="block-title">Keywords</div>
      <p>DataRank, Scientiﬁc dataset, Dataset impact, Citation network</p>
    </div>
  </div>
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#why-measure-dataset-impact" id="toc-why-measure-dataset-impact" class="nav-link" data-scroll-target="#why-measure-dataset-impact"><span class="header-section-number">2</span> Why measure dataset impact?</a></li>
  <li><a href="#materials-and-methods" id="toc-materials-and-methods" class="nav-link" data-scroll-target="#materials-and-methods"><span class="header-section-number">3</span> Materials and methods</a>
  <ul>
  <li><a href="#datasets" id="toc-datasets" class="nav-link" data-scroll-target="#datasets"><span class="header-section-number">3.1</span> Datasets</a>
  <ul class="collapse">
  <li><a href="#opencitations-index-coci" id="toc-opencitations-index-coci" class="nav-link" data-scroll-target="#opencitations-index-coci"><span class="header-section-number">3.1.1</span> OpenCitations Index (COCI)</a></li>
  <li><a href="#microsoft-academic-graph-mag" id="toc-microsoft-academic-graph-mag" class="nav-link" data-scroll-target="#microsoft-academic-graph-mag"><span class="header-section-number">3.1.2</span> Microsoft Academic Graph (MAG)</a></li>
  <li><a href="#pmc-open-access-subset-pmc-oas" id="toc-pmc-open-access-subset-pmc-oas" class="nav-link" data-scroll-target="#pmc-open-access-subset-pmc-oas"><span class="header-section-number">3.1.3</span> PMC Open Access Subset (PMC-OAS)</a></li>
  <li><a href="#genbank" id="toc-genbank" class="nav-link" data-scroll-target="#genbank"><span class="header-section-number">3.1.4</span> GenBank</a></li>
  <li><a href="#figshare" id="toc-figshare" class="nav-link" data-scroll-target="#figshare"><span class="header-section-number">3.1.5</span> Figshare</a></li>
  </ul></li>
  <li><a href="#construction-of-citation-network" id="toc-construction-of-citation-network" class="nav-link" data-scroll-target="#construction-of-citation-network"><span class="header-section-number">3.2</span> Construction of citation network</a></li>
  <li><a href="#visualization-of-citation-network" id="toc-visualization-of-citation-network" class="nav-link" data-scroll-target="#visualization-of-citation-network"><span class="header-section-number">3.3</span> Visualization of citation network</a></li>
  <li><a href="#network-models-for-scientiﬁc-artifacts" id="toc-network-models-for-scientiﬁc-artifacts" class="nav-link" data-scroll-target="#network-models-for-scientiﬁc-artifacts"><span class="header-section-number">3.4</span> Network models for scientiﬁc artifacts</a>
  <ul class="collapse">
  <li><a href="#networkflow" id="toc-networkflow" class="nav-link" data-scroll-target="#networkflow"><span class="header-section-number">3.4.1</span> NetworkFlow</a></li>
  <li><a href="#datarank" id="toc-datarank" class="nav-link" data-scroll-target="#datarank"><span class="header-section-number">3.4.2</span> DataRank</a></li>
  <li><a href="#datarank-fb" id="toc-datarank-fb" class="nav-link" data-scroll-target="#datarank-fb"><span class="header-section-number">3.4.3</span> DataRank-FB</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"><span class="header-section-number">4</span> Results</a></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion"><span class="header-section-number">5</span> Discussion</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">6</span> Conclusion</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="paper-access-btns">
    <a href="./pdf/datarank_wu_bratt_zeng_acuna.pdf" target="_blank" class="btn rounded-0 btn-primary btn-lg page-scroll">Download Preprint PDF</a>
    <a href="https://doi.org/10.1016/j.joi.2020.101013" target="_blank" class="btn rounded-0 btn-lg btn-outline-primary page-scroll">Visit Published DOI</a>
</div>
<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>A citation network is an important source of analysis in science. Citations serve multiple purposes such as crediting an idea, signaling knowledge of the literature, or critiquing others’ work (<code>Martyn, 1975</code>). When citations are thought of as impact, they inform tenure, promotion, and hiring decisions (<code>Meho &amp; Yang, 2007</code>). Furthermore, scientists themselves make decisions based on citations, such as which papers to read and which articles to cite. Citation practices and infrastructures are well-developed for journal articles and conference proceedings. However, there is much less development for dataset citation. This gap affects the increasingly important role that datasets play in scientiﬁc reproducibility (<code>Belter, 2014</code>; <code>On Data Citation Standards &amp; Practices, 2013</code>; <code>Park, You, &amp; Wolfram, 2018</code>; <code>Robinson-García, Jiménez-Contreras, &amp; Torres-Salinas, 2016</code>), where studies use them to conﬁrm or extend the results of other research (<code>Darby et al., 2012</code>; <code>Sieber &amp; Trumbo, 1995</code>). One historical cause of this gap is the difﬁculty in archiving datasets. While less problematic today, the citation practices for datasets take time to develop. Better algorithmic approaches to track dataset usage could improve this state. In this work, we hypothesize that a network ﬂow algorithm could track usage more effectively if it propagates publication and dataset citations differently. With the implementation of this algorithm, then, it will be possible to correct differences in citation behavior between these two types of artifacts, increasing the importance of datasets as ﬁrst class citizens of science.</p>
<p>In this article, we develop a method for assigning credit to datasets from citation networks of publications, assuming that dataset citations have biases. Importantly, our method does not modify the source data for the algorithms. The method does not rely on scientists explicitly citing datasets but infers their usage. We adapt the network ﬂow algorithm of <code>Walker, Xie, Yan, and Maslov (2007)</code> by including two types of nodes: datasets and publications. Our method simulates a random walker that takes into account the differences between obsolescence rates of publications and datasets, and estimates the score of each dataset—the DataRank. We use the metadata from the National Center for Bioinformatics (NCBI) GenBank nucleic acid sequence and Figshare datasets to validate our method. We estimate the relative rank of the datasets with the DataRank algorithm and cross-validate it by predicting the actual usage of them—number of visits to the NCBI dataset web pages and downloads of Figshare datasets. We show that our method is better at predicting both types of usages compared to citations and has other qualitative advantages compared to alternatives. We discuss interpretations of our results and implications for data citation infrastructures and future work.</p>
</section>
<section id="why-measure-dataset-impact" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="why-measure-dataset-impact"><span class="header-section-number">2</span> Why measure dataset impact?</h2>
<p>Scientists may be incentivized to adopt better and broader data sharing behaviors if they, their peers, and institutions are able to measure the impact of datasets (e.g., <code>see Kidwell et al., 2016; Silvello, 2018</code>). In this context, we review impact assess- ment conceptual frameworks and studies of usage statistics and crediting of scientiﬁc works more speciﬁcally. These areas of study aim to develop methods for scientiﬁc indicators of the usage and impact of scholarly outputs. Impact assessment research also derives empirical insights from research products by assessing the dynamics and structures of connections between the outputs. These connections can inform better policy-making for research data management, cyberinfrastructure implementation, and funding allocation.</p>
<p>Methods for measuring usage and impact include a variety of different dimensions of impact, from social media to code use and institutional metrics. Several of these approaches recognize the artiﬁcial distinction between the scientiﬁc process and product (<code>Priem, 2013</code>). For example, altmetrics is one way to measure engagement with diverse research products and to estimate the impact of non-traditional outputs (<code>Priem, 2014</code>). Researchers predict that it will soon become a part of the citation infrastructure to routinely track and value “citations to an online lab notebook, contributions to a software library, bookmarks to datasets from content-sharing sites such as Pinterest and Delicious” (from <code>Priem, 2014</code>). In short, if science has made a difference, it will show up in a multiplicity of places. As such, a correspondingly wider range of metrics are needed to attribute credit to the many locations where research works reﬂect their value. For example, datasets contribute to thousands of papers in NCBI’s Gene Expression Omnibus and these attributions will continue to accumulate, just like paper accumulate citations, for a number of years after the datasets are publicly released (<code>Piwowar, 2013</code>; <code>Piwowar, Vision, &amp; Whitlock, 2011</code>). Efforts to track these other sources of impact include ImpactStory, statistics from FigShare, and Altmetric.com (<code>Robinson-Garcia, Mongeon, Jeng, &amp; Costas, 2017</code>).</p>
<p>Credit attribution efforts include those by federal agencies to expand the deﬁnition of scientiﬁc works that are not pub- lications. For example, in 2013 the National Science Foundation (NSF) recognized the importance of measuring scientiﬁc artifacts other than publications by asking researchers for “products” rather than just publications. This represents a signiﬁ- cant change in how scientists are evaluated (<code>Piwowar, 2013</code>). Datasets, software, and other non-traditional scientiﬁc works are now considered by the NSF as legitimate contributions to the publication record. Furthermore, real-time science is pre- sented in several online mediums; algorithms ﬁlter, rank, and disseminate scholarship as it happens. In sum, the traditional journal article is increasingly being complemented by other scientiﬁc products (<code>Priem, 2013</code>).</p>
<p>Yet despite the crucial role of data in scientiﬁc discovery and innovation, datasets do not get enough credit (<code>Silvello, 2018</code>). If credit was properly accrued, researchers and funding agencies would use this credit to track and justify work and funding to support datasets—consider the recent Rich Context Competition which aimed at to ﬁlling this gap by detecting dataset mentions in full-text papers (<code>Zeng &amp; Acuna, 2020</code>). Because these dataset mentions are not tracked by current citation networks, this leads to biases in dataset citations (<code>Robinson-García et al., 2016</code>). The FAIR (ﬁndable, accessible, interoperable, reproducible) principles of open research data are one major initiative that is spearheading better practices with tracking digital assets such as datasets (<code>Wilkinson et al., 2016</code>). However, the initiative is theoretical, and lacks technical implementation for data usage and impact assessment. There remains a need to establish methods to better estimate dataset usage.</p>
</section>
<section id="materials-and-methods" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="materials-and-methods"><span class="header-section-number">3</span> Materials and methods</h2>
<section id="datasets" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="datasets"><span class="header-section-number">3.1</span> Datasets</h3>
<section id="opencitations-index-coci" class="level4" data-number="3.1.1">
<h4 data-number="3.1.1" class="anchored" data-anchor-id="opencitations-index-coci"><span class="header-section-number">3.1.1</span> OpenCitations Index (COCI)</h4>
<p>The OpenCitations index (COCI) is an index of Crossref’s open DOI-to-DOI citation data. We obtained a snapshot of COCI (November 2018 version), which contains approximately 450 million DOI-to-DOI citations. Speciﬁcally, COCI contains information including citing DOI, cited DOI, the publication date of citing DOI. The algorithm we proposed in the paper requires the publication year. However, not all the DOIs in COCI have a publication date. We will introduce Microsoft Academic Graph to ﬁll this gap.</p>
</section>
<section id="microsoft-academic-graph-mag" class="level4" data-number="3.1.2">
<h4 data-number="3.1.2" class="anchored" data-anchor-id="microsoft-academic-graph-mag"><span class="header-section-number">3.1.2</span> Microsoft Academic Graph (MAG)</h4>
<p>The Microsoft Academic Graph is a large heterogeneous graph consisting of six types of entities: paper, author, institution, venue, event, and ﬁeld of study (<code>Sinha et al., 2015</code>). Concretely, the description of a paper consists of DOI, title, abstract, published year, among other ﬁelds. We downloaded a copy of MAG in November 2019, which contains 208,915,369 papers. As a supplement to COCI, we extract the DOI and published year from MAG to extend those DOIs in COCI without a publication date.</p>
</section>
<section id="pmc-open-access-subset-pmc-oas" class="level4" data-number="3.1.3">
<h4 data-number="3.1.3" class="anchored" data-anchor-id="pmc-open-access-subset-pmc-oas"><span class="header-section-number">3.1.3</span> PMC Open Access Subset (PMC-OAS)</h4>
<p>The PubMed Central Open Access Subset is a collection of full-text open access journal articles in biomedical and life sciences. We obtained a snapshot of PMC-OAS in August 2019. It consists of about 2.5 million full-text articles organized in well-structured XML ﬁles. The articles are identiﬁed by a unique id called PMID. We also obtained a mapping between PMIDs and DOIs from NCBI, which enabled us to integrate PMC-OAS into the citation network.</p>
</section>
<section id="genbank" class="level4" data-number="3.1.4">
<h4 data-number="3.1.4" class="anchored" data-anchor-id="genbank"><span class="header-section-number">3.1.4</span> GenBank</h4>
<p>GenBank is a genetic sequence database that contains an annotated collection of all publicly available nucleotide sequences for almost 420,000 formally described species (<code>Sayers et al., 2019</code>). We obtained a snapshot of the GenBank database (version 230) with 212,260,377 gene sequences. We remove those sequences without submission date. This left us with 77,149,105 sequences.</p>
</section>
<section id="figshare" class="level4" data-number="3.1.5">
<h4 data-number="3.1.5" class="anchored" data-anchor-id="figshare"><span class="header-section-number">3.1.5</span> Figshare</h4>
<p>Figshare is a multidisciplinary, open access research repository where researchers can deposit and share their research output. Figshare allows users to upload various formats of research output, including ﬁgures, datasets, and other media (<code>Thelwall &amp; Kousha, 2016</code>). In order to encourage data sharing, all research data made publicly available has unlimited storage space and is allocated a DOI. This DOI is used by scientists to cite Figshare resources using traditional citation methods. Figshare makes the research data publicly and permanently available which mitigates the resource decay problem and improves research reproducibility (<code>Zeng, Shema, &amp; Acuna, 2019</code>). A Figshare DOI contains the string ‘ﬁgshare’ (e.g., ‘10.6084/m9.ﬁgshare.6741260’). We can leverage this feature to determine whether a publication cites Figshare resources by detecting Figshare-generated DOIs.</p>
</section>
</section>
<section id="construction-of-citation-network" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="construction-of-citation-network"><span class="header-section-number">3.2</span> Construction of citation network</h3>
<p>The citation networks in this paper consist of two types of nodes and two types of edges. The node is represented by the paper and the dataset and the edge is represented by the citation between two nodes. Concretely, papers cite each other to form paper-paper edges as datasets can only be cited by papers which are represented by the paper-dataset edges. As shown in the construction workﬂow (Fig. 2), we build the paper-paper citation network using COCI and MAG and build two separate paper-dataset edge sets using GenBank and Figshare. Then we integrate the paper-dataset edge sets into the paper-paper citation network to form two complete citation networks. The construction workﬂow is illustrated in Fig. 2.</p>
<p>Please read the full paper for details of data processing.</p>
</section>
<section id="visualization-of-citation-network" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="visualization-of-citation-network"><span class="header-section-number">3.3</span> Visualization of citation network</h3>
<div id="fig-network" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-network-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/f1.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-network-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Visualization of publication-dataset citation network. Papers appear as red nodes and Datasets are green. The lighter the color, the older the resource. (For interpretation of the references to color in this ﬁgure legend, the reader is referred to the web version of this article.)
</figcaption>
</figure>
</div>
</section>
<section id="network-models-for-scientiﬁc-artifacts" class="level3" data-number="3.4">
<h3 data-number="3.4" class="anchored" data-anchor-id="network-models-for-scientiﬁc-artifacts"><span class="header-section-number">3.4</span> Network models for scientiﬁc artifacts</h3>
<section id="networkflow" class="level4" data-number="3.4.1">
<h4 data-number="3.4.1" class="anchored" data-anchor-id="networkflow"><span class="header-section-number">3.4.1</span> NetworkFlow</h4>
<p>We adapt the network model proposed in <code>Walker et al. (2007)</code>. This method, which we call NetworkFlow here, is inspired by PageRank and addresses the issue that citation networks are always directed back in time. In this model, each vertex in the graph is a publication. The goal of this method is to propagate a vertex’s impact through its citations.</p>
</section>
<section id="datarank" class="level4" data-number="3.4.2">
<h4 data-number="3.4.2" class="anchored" data-anchor-id="datarank"><span class="header-section-number">3.4.2</span> DataRank</h4>
<p>In this article, we extend NetworkFlow to accommodate different kinds of nodes. The extension considers that the probability of starting at any single node should depend on whether the node is a publication or a dataset. This is, publications and datasets may vary in their relevance period. We call this new algorithm DataRank. Mathematically, we redeﬁne the starting probability in <a href="#eq-datarank" class="quarto-xref">Equation&nbsp;1</a> of the ith node as:</p>
<p><span id="eq-datarank"><span class="math display">
\begin{equation}
\rho_{i}^{\text{DataRank}}=\begin{cases}
\exp\left(-\frac{\text{age}_{i}}{\tau_{\text{pub}}}\right) &amp; \text{if } i \text{ is a publication}\\
\exp\left(-\frac{\text{age}_{i}}{\tau_{\text{dataset}}}\right) &amp; \text{if } i \text{ is a dataset}
\end{cases}
\end{equation}
\tag{1}</span></span></p>
</section>
<section id="datarank-fb" class="level4" data-number="3.4.3">
<h4 data-number="3.4.3" class="anchored" data-anchor-id="datarank-fb"><span class="header-section-number">3.4.3</span> DataRank-FB</h4>
<p>In DataRank, each time the walker moves, there are two options: to stop with a probability <span class="math inline">\alpha</span> or to continue the search through the reference list with a probability <span class="math inline">1-\alpha</span>. However, there may exist a third choice: to continue the search through papers who cite the current paper. In other words, the researcher can move in two directions: forwards and backwards. We call this modified method DataRank-FB. In this method, one may stop with a probability <span class="math inline">\alpha-\beta</span>, continue the search forward with a probability <span class="math inline">1-\alpha</span>, and backward with a probability <span class="math inline">\beta</span>. To keep them within the unity simplex, the parameters must satisfy <span class="math inline">\alpha&gt;0</span>, <span class="math inline">\beta&gt;0</span>, and <span class="math inline">\alpha&gt;\beta</span>,</p>
<p>Then, we deﬁne another transition matrix from the citation network as follows <span id="eq-transition-probability-1"><span class="math display">
M_{ij}=\begin{cases}
\frac{1}{k_{j}^{\text{in}}} &amp; \text{if } j  \text{ cited by } i \\
0 &amp; \text{o.w.}
\end{cases}
\tag{2}</span></span></p>
<p>where <span class="math inline">k_{j}^{\text{in}}</span> is the number of papers that cite <span class="math inline">j</span>.</p>
<p>We update the average path length to all papers in the network starting from <span class="math inline">\rho</span> as</p>
<p><span id="eq-average-path-length-1"><span class="math display">
\begin{equation}
T=I\cdot\rho+(1-\alpha)W\cdot\rho+\beta M\cdot\rho+(1-\alpha)^{2}W^{2}\rho+\beta^{2}M^{2}\rho+\cdots
\end{equation}
\tag{3}</span></span></p>
</section>
</section>
</section>
<section id="results" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="results"><span class="header-section-number">4</span> Results</h2>
<p>We aim at ﬁnding whether the estimation of the rank of a dataset based on citation data is related to a real-world measure of relevance such as page views or downloads. We propose a method for estimating rankings that we call DataRank, which considers differences in citation dynamics for publications and datasets. We also propose some variants to this method, and compare all of them to standard ranking algorithms. We use the data of visits of GenBank datasets and downloads of Figshare datasets as measure of real usage. Thus, we will investigate which of the methods work best for ranking them.</p>
<p>Please read the full paper for details.</p>
</section>
<section id="discussion" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="discussion"><span class="header-section-number">5</span> Discussion</h2>
<p>The goal of this article is to better evaluate the importance of datasets through article citation network analysis. Compared with the mature citation mechanisms of articles, referencing datasets is still in its infancy. Acknowledging the long time the practice of citing datasets will take to be adopted, our research aims at recovering the true importance of datasets even if their citations are biased compared to publications.</p>
<p>Scholars disagree on how to give credit to research outcomes. Regardless of how disputed citations are as a measure of credit, they complement other measures that are harder to quantify such as peer review assessment or real usage such as downloads. Citations, however, are rarely used for datasets (<code>Altman &amp; Crosas, 2013)</code>, giving these important research outcomes less credit that they might deserve. Our proposal aims at solving some of these issues by constructing a network ﬂow that is able to successfully predict real usage better than other methods. While citations are not a perfect measure of credit, having a method that can at least attempt to predict usage is advantageous.</p>
<p>Previous research has examined ways of normalizing citations by time, ﬁeld, and quality. This relates to our DataRank algorithm in that we are trying to normalize the citations by year and artifact type.</p>
<p>We ﬁnd it useful to interpret the effect of different decay times on the performance of DataRank.</p>
</section>
<section id="conclusion" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="conclusion"><span class="header-section-number">6</span> Conclusion</h2>
<p>Understanding how datasets are used is an important topic in science. Datasets are becoming crucial for the reproduction of results and the acceleration of scientiﬁc discoveries. Scientists, however, tend to dismiss citing datasets and therefore there is no proper measurement of how impactful datasets are. Our method uses the publication-publication citation network to propagate the impact to the publication-dataset citation network. Using two databases of real dataset networks, we demonstrate how our method is able to predict actual usage more accurately than other methods. Our results suggest that datasets have different citation dynamics to those of publications. In sum, our study provides a prescriptive model to understand how citations interact in publication and dataset citation networks and gives a concrete method for producing ranks that are predictive of actual usage.</p>
<p>Our study advances an investigation of datasets and publications with novel ideas from network analysis. Our work puts together concepts from other popular network ﬂow estimation algorithms such as PageRank (<code>Brin &amp; Page, 1998</code>) and NetworkFlow (<code>Walker et al., 2007</code>). While scientists might potentially take a great amount of time to change their cita- tion behavior, we could use techniques such as the one put forth in this article to accelerate the credit assignment for datasets. Ultimately, the need for tracking datasets will only become more pressing and therefore we must adapt or miss the opportunity to make datasets ﬁrst-class citizens of science.</p>


</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" id="quarto-citation"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@article{zeng2020,
  author = {Zeng, Tong and Wu, Longfeng and Bratt, Sarah and E Acuna,
    Daniel},
  title = {Assigning Credit to Scientific Datasets Using Article
    Citation Networks},
  journal = {Journal of Informetrics},
  volume = {14},
  number = {2},
  pages = {101013},
  date = {2020-05},
  url = {https://www.sciencedirect.com/science/article/pii/S1751157719301841},
  doi = {10.1016/j.joi.2020.101013},
  issn = {1751-1577},
  langid = {en},
  abstract = {A citation is a well-established mechanism for connecting
    scientiﬁc artifacts. Citation networks are used by citation analysis
    for a variety of reasons, prominently to give credit to scientists’
    work. However, because of current citation practices, scientists
    tend to cite only publications, leaving out other types of artifacts
    such as datasets. Datasets then do not get appropriate credit even
    though they are increasingly reused and experimented with. We
    develop a network ﬂow measure, called DataRank, aimed at solving
    this gap. DataRank assigns a relative value to each node in the
    network based on how citations ﬂow through the graph,
    differentiating publication and dataset ﬂow rates. We evaluate the
    quality of DataRank by estimating its accuracy at predicting the
    usage of real datasets: web visits to GenBank and downloads of
    Figshare datasets. We show that DataRank is better at predicting
    this usage compared to alternatives while offering additional
    interpretable outcomes. We discuss improvements to citation behavior
    and algorithms to properly track and assign credit to datasets.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-zeng2020" class="csl-entry quarto-appendix-citeas" role="listitem">
Zeng, Tong, Longfeng Wu, Sarah Bratt, and Daniel E Acuna. 2020.
<span>“Assigning Credit to Scientific Datasets Using Article Citation
Networks.”</span> <em>Journal of Informetrics</em> 14 (2): 101013. <a href="https://doi.org/10.1016/j.joi.2020.101013">https://doi.org/10.1016/j.joi.2020.101013</a>.
</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/tong-zeng\.github\.io\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Powered by <a href="https://quarto.org/" target="_blank">Quarto</a></p>
</div>   
    <div class="nav-footer-center">
<p>Copyright © 2024 TONG ZENG. ALL RIGHTS RESERVED.</p>
</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>